certora/specs/sharesAssetsConversion.spec

```diff
@@ -0,0 +1,257 @@
+
+    assert jointAmount >= amount1 + amount2, "Found advantage in combining accounts";
+    assert jointAmount < amount1 + amount2 + 2, "Found advantage in splitting accounts";
+    // The following assertion fails (as expected):
+    // assert jointAmount < amount1 + amount2 + 1, "Found advantage in splitting accounts";
+}
+
+// Similar rule as above for shares
+rule convertSumOfAssetsPreserved(uint256 assets1, uint256 assets2) {
+    uint256 shares1 = convertToShares(assets1);
+    uint256 shares2 = convertToShares(assets2);
+    uint256 sumAssets = assets1 + assets2;
+	require sumAssets >= assets1 + assets2; // Prevent overflow
+    uint256 jointShares = convertToShares(sumAssets);
+
+	assert jointShares >= shares1 + shares2, "Convert sum of assets bigger than parts";
+	assert jointShares < shares1 + shares2 + 2, "Convert sum of assets far smaller than parts";
+}
+
+// Ensure that maxWithdraw and maxRedeem are in line with the conversion functions.
+rule maxWithdrawRedeemCompliance(address owner) {
+    uint256 shares = balanceOf(owner);
+    uint256 amountConverted = convertToAssets(shares);
+
+    assert maxWithdraw(owner) <= amountConverted, "Can withdraw more than converted amount";
+    assert maxRedeem(owner) <= shares, "Can redeem more than available shares)";
+}
+
+
+// Ensure that previewWithdraw and previewRedeem are in line with the conversion functions.
+rule previewWithdrawRedeemCompliance(uint256 value) {
+    env e;
+    uint256 assets = convertToAssets(value);
+    uint256 shares = convertToShares(value);
+
+    assert previewWithdraw(e, value) >= shares, "Preview withdraw takes less shares than converted";
+    assert previewRedeem(e, value) <= assets, "Preview redeem yields more assets than converted";
+
+	// The following rule protects the client.
+    assert previewWithdraw(e, value) <= shares + 1, "Preview withdraw costs too many shares";
+}
+
+
+/* From ERC4626:
+ * "previewWithdraw ... MUST return as close to and no fewer than the exact amount of Vault 
+ *  shares that would be burned in a withdraw call in the same transaction. I.e. withdraw
+ *  should return the same or fewer shares as previewWithdraw if called in the same
+ *  transaction."
+ */
+rule previewWithdrawNearlyWithdraw(uint256 assets) {
+    env e;
+	address owner = e.msg.sender;  // Handy alias
+	uint256 previewShares = previewWithdraw(e, assets);
+	uint256 withdrawShares = withdraw(e, assets, owner, owner);
+
+	assert withdrawShares <= previewShares, "Withdraw returns more shares than preview";
+	assert withdrawShares + 1 >= previewShares, "Withdraw returns far less shares than preview";
+}
+
+
+/* From ERC4626:
+ * "previewRedeem ... MUST return as close to and no more than the exact amount
+ *  of assets that would be withdrawn in a redeem call in the same transaction.
+ *  I.e. redeem should return the same or more assets as previewRedeem if called
+ *  in the same transaction."
+ */
+rule previewRedeemNearlyRedeem(uint256 shares) {
```

Comment:
Same question as on previous rule - how is this different than `previewWithdrawRedeemCompliance`?
in this case you showed both sides of the range, so it's slightly better imo